### ISSUE TITLE
Reproduce the MathWriting baseline

### ISSUE THREAD

**Issue by oksr** [oksr/open-math-notes#5]
## Question

Train the CTC Transformer to the paper's numbers (CER 5.49 / 60% exact on MathWriting test; valid 4.52 / 71%) using the pipeline from [Training pipeline and evaluation harness](18-training-pipeline-and-eval-harness.md), on MathWriting train + synthetic **plus CROHME** InkML. Budget: **three full runs** — each run either overnight on the M5 Pro or on a cloud A100 within the $200 ceiling, chosen from the measured samples/sec (HITL: the user approves any cloud spend). Between runs, sweep at small scale on the Mac.

Report per run: MathWriting test exact / CER / within-1-edit, CROHME 2019 test exact, the 50 held-out personal samples if recorded, the per-construct error analysis on v1 notation, wall-clock and cost. If three runs miss the paper's numbers, say so plainly — that triggers the image fallback in [Architecture confirmation](12-architecture-confirmation.md). Resolve with the best checkpoint's location and numbers.

**Comment by oksr** [comment]
## Progress (session 2026-09-01) — run 1 is training on the Mac; ticket stays open

**Status.** Run 1 (paper recipe, full 100k steps, MathWriting train + synthetic + CROHME train, bf16 on MPS) started at ~23:50 on 2026-09-01 on the M5 Pro and is healthy: 280–320 samples/s steady, allocator memory flat at 10.4 GB (13 GB in `top`), loss 7.48 → 4.12 over the first 150 steps, no non-finite batches. **ETA ≈ 24 h** for all 100k steps; validation (2,048 valid inks, greedy) every 2k steps, `best.pt` kept at the best validation CER, resumable from `last.pt` every 1k steps.

```sh
tail -f ml/runs/run1/stdout.log                       # progress; "valid @ N" lines are the checkpoints
pgrep -fl "scripts/train.py --name run1"              # pid;  kill it to stop — rerunning the same command resumes
uv run python scripts/evaluate.py runs/run1/best.pt   # when done (or any time): mw-test / mw-valid / crohme-test2019, per-construct table, confusions
```

**What landed** (commits `7662f75`, `d261d12`, `4078b74`, `a1a3457` on `main`, 40 tests):
- `scripts/train.py`: full-run loop — multi-split loader, length-bucketed batches, checkpoints + mid-epoch resume, periodic validation, LR schedules, bf16 autocast, JSONL log, W&B optional. `scripts/evaluate.py`: headline metrics vs the paper, **per-construct breakdown on v1 notation** (fraction / root / sum / integral / limit / derivative / greek / sub / superscript / logic / relation / function name; matrix + multi-line flagged post-v1), top token confusions. `scripts/cloud/{push,bootstrap,launch,pull}.sh` + runbook **`ml/docs/cloud-run.md`**.
- All splits preprocessed (bezier, ~0.9 GB of shards). Training set = 622,095 inks (mw-train 217,819 + mw-synthetic 392,424 + crohme-train 11,852). Evaluation shards now **keep the ctc-too-short inks** (1–3% of valid/test), so every ink of mw-test (7,644), mw-valid (15,674) and crohme-test2019 (1,199) is scored — the caveat from ticket 18 is closed.

**Measured on the M5 Pro (bezier, batch 256, 35M params)**

| config | samples/s | 100k steps |
|---|---|---|
| random batches, pad 64 (ticket 18's number) | 97–114 | ~70 h |
| length-bucketed batches | 207 burst | ~34 h |
| + bf16 autocast on MPS (`--mps-bf16`) | **~290 sustained** | **~24 h** |

bf16's loss curve matched fp32 step for step over 400 steps (valid CER 79.9 vs 80.3 at step 400 on a 30k subset — same thing). Random batches waste 77% of padded compute (median 28 curves, batches padded to 128); bucketing fixes that.

**Three things that broke the first three attempts, all fixed and documented in the runbook:**
1. **NaN at step ~100** — 49 CROHME train inks carried NaN/inf in the Bézier d1/d2 features (`min(nan, 10.0)` is nan in Python). Featurizer is NaN-safe now, CROHME re-sharded, and the loop skips + counts any non-finite batch instead of poisoning the weights.
2. **35–41 GB process memory → swap storm, throughput ÷3–5** — the MPS caching allocator hoards freed blocks per shape; bucketing packs the longest inks (synthetic runs to 300 curves) into batches with several times the activations; and `driver_allocated_memory()` under-reports the real footprint. Fix: each 256-ink batch runs in chunks of ≤16,384 padded steps with gradient accumulation (recipe's batch 256 holds exactly — a test checks the gradient is identical) and the cache is released every step (4% throughput). Result: ~8–13 GB.
3. Watermark ratios (`PYTORCH_MPS_*_WATERMARK_RATIO`) are not a fix; they turn growth into OOM.

**Decisions for you (HITL):**
- **Mac vs A100 for this run.** The Mac run is free and needs ~24 h with the laptop on AC, lid open (it resumes after a sleep or a kill). The A100 path is ready end to end (`cloud-run.md`, ~$4–14/run on RunPod Secure A100 at $1.39/h, measured on the box by `bootstrap.sh` before any long spend) but needs your RunPod account + SSH key, and your go-ahead on the spend. My recommendation: let the Mac run finish; keep the A100 for runs 2–3 if the sweep between runs needs several candidates.
- **Memory on the Mac.** When run 1 started, the OS already had ~21 GB in swap from other apps (browsers, several `claude.exe`, `stable`). The run now fits in ~13 GB, but closing memory-heavy apps overnight keeps it at full speed.
- **Personal set.** The 50 held-out personal samples are not recorded (InkCapture has not run on the iPad yet), so run 1's report will cover MathWriting test/valid and CROHME 2019 only.

**Next session on this ticket:** read `runs/run1/stdout.log` (or `report.json`), run `evaluate.py` on `best.pt`, post the run-1 numbers vs the paper (test CER 5.49 / exact 60% / within-1 72%; valid 4.52 / 71 / 81) with the per-construct table, then choose run 2's sweep. Candidates from what I saw in the data: the Bézier time coefficients (γ1–γ3) are heavy-tailed (>10 in ~14% of inks, max ≈440; the model has no input normalization) → input standardization or a γ clamp; cosine LR decay; the synthetic ablation only if run 1 is far off.

**Comment by oksr** [comment]
**First validation point (run 1, step 2,000 of 100k, 00:27 on 2026-09-02):** greedy on 2,048 valid inks — **CER 25.3, exact 14.8%, within-1-edit 30.2%**; training loss 7.48 → 0.86. The loop learns. Throughput steady at ~290 samples/s, allocator memory flat at 10.4 GB, `best.pt` and `last.pt` written. Nothing else to do until the run is further along; the next validation lands every 2k steps (~30 min each) in `ml/runs/run1/stdout.log`.
